### DISCUSSION TITLE
sincos and thread safety

### DISCUSSION THREAD

**Discussion by ghalimi** [stdlib-js/stdlib#382] [Q&A] [answered]
What does the comment `WARNING: not thread safe` mean on this [line](https://github.com/stdlib-js/stdlib/blob/develop/lib/node_modules/%40stdlib/math/base/special/bessely1/lib/y1.js#L63) in relation to the `sincos` function? Since JavaScript is single-threaded (taking web workers aside), how could thread safety be an issue? Is this a legacy of an anterior C implementation?

**Comment by kgryte** [comment, marked as answer]
This means that, if someone was porting to a language supporting multiple threads (e.g., C), then porting as is would not be thread safe. In JavaScript, we can elevate memory allocations out of functions as code does not need be reentrant. In C, we would not be able to do so.

**Reply by trusktr** [reply]
If you fork a C program, then both programs have their own copy of the in-scope variables right? Wouldn't it still be thread safe then?

But in any case, I think the comments aren't needed. Someone porting to any other language should know what they're doing in their target language.

I believe the comments will cause more confusion there than benefit.

**Reply by kgryte** [reply]
Re: fork. Yes, that is correct.

The source comments are useful for us--that's why they are there--serving as a reminder as to how our internal native C and JS implementations should differ.

**Reply by kgryte** [reply]
New discovery today: JavaScript code which is **not** thread safe can affect whether V8 is able to perform certain optimizations. If a function is not thread safe, this can trigger de-optimizations. This was discovered on newer versions of V8 during profiling.
